src/main/java/io/confluent/idesidecar/restapi/messageviewer/RecordDeserializer.java

```diff
@@ -224,16 +245,65 @@ public DecodedResult deserialize(
           .builder()
           .value(deserializedData)
           .build();
-    } catch (RestClientException e) {
-      cacheSchemaFetchError(e, schemaId);
-      return new DecodedResult(onFailure(encoderOnFailure, bytes), e.getMessage());
-    } catch (IOException | SerializationException e) {
-      var data = onFailure(encoderOnFailure, bytes);
-      Log.error("Error deserializing: %s", data, e);
-      return new DecodedResult(data, e.getMessage());
+    } catch (Exception e) {
+      if (e instanceof RestClientException || isNetworkRelatedException(e)) {
+        // IMPORTANT: We must cache RestClientException, and more importantly,
+        //            network-related IOExceptions, to prevent the sidecar from
+        //            bombarding the Schema Registry servers with requests for every
+        //            consumed message when we encounter a schema fetch error.
+        cacheSchemaFetchError(e, schemaId);
+        return new DecodedResult(onFailure(encoderOnFailure, bytes), e.getMessage());
+      } else if (
+          e instanceof SerializationException
+              || e instanceof JsonProcessingException
+      ) {
+        // We don't cache SerializationException because it's not a schema fetch error
+        // but rather a deserialization error, scoped to the specific message.
+        return new DecodedResult(onFailure(encoderOnFailure, bytes), e.getMessage());
+      }
+      // If we reach this point, we have an unexpected exception, so we rethrow it.
+      throw new RuntimeException("Failed to deserialize record", e);
     }
   }
 
+  private boolean isRetryableException(Throwable throwable) {
+    return (
+        isNetworkRelatedException(throwable) ||
+            (throwable instanceof RestClientException
+                && isRestClientExceptionRetryable((RestClientException) throwable)
+    ));
+  }
+
+  private boolean isNetworkRelatedException(Throwable throwable) {
+    var e = (Exception) throwable;
+    return (
+        e instanceof ConnectException
+            || e instanceof SocketTimeoutException
+            || e instanceof UnknownHostException
+            || e instanceof UnknownServiceException
```

Comment:
These three were arbitrarily chosen, what matters most is that `ConnectException` is part of this.